apps/mosaico/src/Article/Article.purs

```diff
@@ -392,3 +396,33 @@ render { props, state, setState } =
         -- Does the string start with wanted element
         isElem elemName elemString =
           Just 0 == String.indexOf (Pattern elemName) elemString
+        
+
+    insertAdsIntoBodyText :: String -> String -> Array BodyElement -> Array BodyElement  
+    insertAdsIntoBodyText contentUnit1 contentUnit2 body =
+      let ad1 = Ad contentUnit1
+          ad2 = Ad contentUnit2
+      in if elements > 15
+        then fromMaybe body $ do
+          bodyWithAd <- insertAt (findAdSpace body $ elements/3) ad1 body
+          insertAt (findAdSpace bodyWithAd $ 2 * elements/3) ad2 bodyWithAd
+        else if elements > 6
+          then fromMaybe body $ insertAt (findAdSpace body $ elements/2) ad1 body
+          else body `snoc` ad1
+      where
+        elements = length body
+        findAdSpace :: Array BodyElement -> Int -> Int
+        findAdSpace body' i 
+          | i > elements = elements
+          | otherwise    = case body' !! (i-1) of
+            Just (Html _) -> case body' !! i of
+              Just (Html _) -> i
+              _             -> findAdSpace body' (i+1)
+            _             -> findAdSpace body' (i+1)
+
+    removeBodyErrors :: Array (Either String BodyElement) -> Array BodyElement
+    removeBodyErrors body = 
+      let partitioned = partition isRight body
+      in removeEither <$> partitioned.yes
+      where
+        removeEither = fromRight (Html "")
```

Comment:
By first looking at this I thought it turns every `Left String` into empty `Html` blocks but by reading the entire function, that isn't the case, right? 😄

I'm personally annoyed every time I need to do the same thing, as there are no `takeRights` or `takeLefts` functions easily available (in Data.Either or similar). As this is a recurring problem, we should probably consider using something like [Data.Filterable](https://pursuit.purescript.org/packages/purescript-filterable/4.0.1/docs/Data.Filterable#v:partitionMap). No need to do it in this PR though.

Anyway, the solution is fine as the empty `Html` blocks should never get created, but I think there are better ways of solving this with for instance `foldl`. Want to give it a try as an exercise? 🙂

Reply:
Okay sure. Best as part of this PR or a later one?